contributors/devel/sig-architecture/api_changes.md

```diff
@@ -966,7 +966,7 @@ cluster; removing the upgradability concern allows developers to make rapid
 progress; in particular, API versions can increment faster than the minor
 release cadence and the developer need not maintain multiple versions;
 developers should still increment the API version when object schema or
-semantics change in an [incompatible way](#on-compatibility)
+semantics change in an [incompatible way](#on-compatibility). Users should not expect that alpha objects or fields will behave the same during or after a rolling upgrade to a version of Kubernetes with the object or field enabled as beta.
```

Comment:
"with the" -> "where the"?

Reply:
While I agree with Tim that we should discourage unnecessary breakage during alpha phase, my main concern is making sure users that never enabled an alpha feature/api/field are _never_ broken when the feature graduates to beta.

For that reason, I think there's more freedom around alpha *versions* than alpha *fields*.

An alpha *version* is not enabled by default, and a server without the alpha *version* enabled will never even try to read persisted data from etcd into those alpha types.

Persisted data in an alpha *field* in a beta/ga version is read into the go types and validated even if the alpha feature is disabled. That means if you make a breaking change to the validation or semantics AND promote to beta in the same release, the n-1 API server that only knows the previous alpha logic can break.

I think if you make a breaking validation/semantic change to a field while in alpha, you have to wait a release to promote to beta so the v1.(n-1) alpha-and-disabled-by-default-but-still-exposed-to-persisted-data server is coherent with the v1.n beta-and-enabled-by-default server.